File: scripts/download_sim8_player.js
#!/usr/bin/env node

const fs = require("fs");
const https = require("https");
const path = require("path");

const [inputPath, outputDir, targetPlayer, fileSlug] = process.argv.slice(2);
const tournament = "Simulation-8 / 模拟赛-8";
const tournamentId = "6a1d3d1258ebe27b197a8cca";

if (!inputPath || !outputDir || !targetPlayer || !fileSlug) {
  console.error(
    "Usage: node scripts/download_sim8_player.js <ids.json> <output_dir> <target_player> <file_slug>"
  );
  process.exit(1);
}

fs.mkdirSync(outputDir, { recursive: true });
const ids = JSON.parse(fs.readFileSync(inputPath, "utf8"));
const selected = [];
const failed = [];
let completed = 0;

function fetchJson(url, retries = 3) {
  return new Promise((resolve, reject) => {
    const request = https.get(url, { timeout: 30000 }, (response) => {
      if (response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        response.resume();
        resolve(fetchJson(new URL(response.headers.location, url).href, retries));
        return;
      }

      let body = "";
      response.setEncoding("utf8");
      response.on("data", (chunk) => {
        body += chunk;
      });
      response.on("end", () => {
        try {
          resolve(JSON.parse(body));
        } catch (error) {
          reject(new Error(`Invalid JSON from ${url}: ${error.message}`));
        }
      });
    });

    request.on("timeout", () => request.destroy(new Error("Request timed out")));
    request.on("error", (error) => {
      if (retries > 1) {
        setTimeout(() => resolve(fetchJson(url, retries - 1)), 1000);
      } else {
        reject(error);
      }
    });
  });
}

async function main() {
  async function processMatch(matchId) {
    try {
      const litePath = path.join(outputDir, `${matchId}_lite.json`);
      const data = fs.existsSync(litePath)
        ? JSON.parse(fs.readFileSync(litePath, "utf8"))
        : await fetchJson(`https://www.botzone.org.cn/match/${matchId}?lite=true`);
      const playerNames = (data.players || []).map((player) => player.name);

      if (playerNames.includes(targetPlayer)) {
        fs.writeFileSync(litePath, JSON.stringify(data, null, 2));
        fs.writeFileSync(
          path.join(outputDir, `${matchId}_${fileSlug}_full_log.json`),
          JSON.stringify(data.logs, null, 2)
        );
        fs.writeFileSync(
          path.join(outputDir, `${matchId}_metadata.json`),
          JSON.stringify(
            {
              matchId,
              tournament,
              tournamentId,
              targetPlayer,
              status: data.status,
              success: data.success,
              players: data.players,
              initdata: data.initdata ? JSON.parse(data.initdata) : null,
              viewurl: data.viewurl,
              logRecords: Array.isArray(data.logs) ? data.logs.length : null,
              collectedAt: new Date().toISOString()
            },
            null,
            2
          )
        );
        selected.push(matchId);
      }
    } catch (error) {
      failed.push({ matchId, error: error.message });
      console.error(`${matchId} failed: ${error.message}`);
    } finally {
      completed++;
      console.log(`${completed}/${ids.length} ${matchId} selected=${selected.length}`);
    }
  }

  let nextIndex = 0;
  async function worker() {
    while (nextIndex < ids.length) {
      const index = nextIndex++;
      await processMatch(ids[index]);
    }
  }

  await Promise.all(Array.from({ length: 8 }, () => worker()));
  selected.sort();

  fs.writeFileSync(
    path.join(outputDir, "manifest.json"),
    JSON.stringify(
      {
        tournament,
        tournamentId,
        targetPlayer,
        fileSlug,
        scannedMatchCount: ids.length,
        selectedMatchCount: selected.length,
        selectedMatchIds: selected,
        failures: failed,
        collectedAt: new Date().toISOString()
      },
      null,
      2
    )
  );

  console.log(JSON.stringify({ scanned: ids.length, selected: selected.length, failed: failed.length }));
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/download_sim8_player.py
#!/usr/bin/env python3

import argparse
import concurrent.futures
import json
import time
import urllib.request
from pathlib import Path


TOURNAMENT = "Simulation-8 / 模拟赛-8"
TOURNAMENT_ID = "6a1d3d1258ebe27b197a8cca"


def fetch_json(url, retries=4):
    last_error = None
    for attempt in range(retries):
        try:
            request = urllib.request.Request(
                url,
                headers={
                    "Accept": "application/json",
                    "Accept-Encoding": "identity",
                    "User-Agent": "Mozilla/5.0 Botzone Sim8 player collector",
                },
            )
            with urllib.request.urlopen(request, timeout=60) as response:
                return json.loads(response.read().decode("utf-8"))
        except Exception as error:
            last_error = error
            time.sleep(1 + attempt * 2)
    raise last_error


def write_json(path, data):
    temp_path = path.with_suffix(path.suffix + ".tmp")
    temp_path.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")
    temp_path.replace(path)


def process_match(output_dir, target_player, file_slug, match_id):
    lite_path = output_dir / f"{match_id}_lite.json"
    data = (
        json.loads(lite_path.read_text(encoding="utf-8"))
        if lite_path.exists()
        else fetch_json(f"https://www.botzone.org.cn/match/{match_id}?lite=true")
    )
    if target_player not in [player.get("name") for player in data.get("players", [])]:
        return match_id, False

    write_json(lite_path, data)
    write_json(output_dir / f"{match_id}_{file_slug}_full_log.json", data.get("logs"))
    initdata = data.get("initdata")
    if isinstance(initdata, str):
        try:
            initdata = json.loads(initdata)
        except json.JSONDecodeError:
            pass
    write_json(
        output_dir / f"{match_id}_metadata.json",
        {
            "matchId": match_id,
            "tournament": TOURNAMENT,
            "tournamentId": TOURNAMENT_ID,
            "targetPlayer": target_player,
            "status": data.get("status"),
            "success": data.get("success"),
            "players": data.get("players"),
            "initdata": initdata,
            "viewurl": data.get("viewurl"),
            "logRecords": len(data["logs"]) if isinstance(data.get("logs"), list) else None,
            "collectedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        },
    )
    return match_id, True


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("ids_json")
    parser.add_argument("output_dir")
    parser.add_argument("target_player")
    parser.add_argument("file_slug")
    parser.add_argument("--workers", type=int, default=8)
    args = parser.parse_args()

    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)
    match_ids = json.loads(Path(args.ids_json).read_text(encoding="utf-8"))
    selected = []
    failures = []

    with concurrent.futures.ThreadPoolExecutor(max_workers=args.workers) as executor:
        futures = {
            executor.submit(
                process_match, output_dir, args.target_player, args.file_slug, match_id
            ): match_id
            for match_id in match_ids
        }
        completed = 0
        for future in concurrent.futures.as_completed(futures):
            match_id = futures[future]
            completed += 1
            try:
                _, included = future.result()
                if included:
                    selected.append(match_id)
                print(
                    f"{completed}/{len(match_ids)} {match_id} selected={len(selected)}",
                    flush=True,
                )
            except Exception as error:
                failures.append({"matchId": match_id, "error": str(error)})
                print(f"{completed}/{len(match_ids)} {match_id} failed: {error}", flush=True)

    selected.sort()
    write_json(
        output_dir / "manifest.json",
        {
            "tournament": TOURNAMENT,
            "tournamentId": TOURNAMENT_ID,
            "targetPlayer": args.target_player,
            "fileSlug": args.file_slug,
            "scannedMatchCount": len(match_ids),
            "selectedMatchCount": len(selected),
            "selectedMatchIds": selected,
            "failures": failures,
            "collectedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        },
    )
    print(
        json.dumps(
            {"scanned": len(match_ids), "selected": len(selected), "failed": len(failures)}
        ),
        flush=True,
    )


if __name__ == "__main__":
    main()
